consensus/clique/src/main/java/tech/pegasys/pantheon/consensus/clique/VoteTallyCache.java

```diff
@@ -80,24 +92,22 @@ private VoteTally populateCacheUptoAndIncluding(final BlockHeader start) {
     return constructMissingCacheEntries(intermediateBlocks, voteTally);
   }
 
-  private VoteTally findMostRecentAvailableVoteTally(
-      final BlockHeader header, final Deque<BlockHeader> intermediateBlockHeaders) {
-    intermediateBlockHeaders.push(header);
-    VoteTally voteTally = voteTallyCache.getIfPresent(header.getParentHash());
-    if ((voteTally == null) && (epochManager.isEpochBlock(header.getNumber()))) {
-      final CliqueExtraData extraData = CliqueExtraData.decode(header.getExtraData());
-      voteTally = new VoteTally(extraData.getValidators());
+  private VoteTally getValidatorsAfter(final BlockHeader header) {
+    if (epochManager.isEpochBlock(header.getNumber())) {
+      final CliqueBlockInterface blockInterface = new CliqueBlockInterface();
+      return new VoteTally(blockInterface.validatorsInBlock(header));
     }
 
-    return voteTally;
+    return voteTallyCache.getIfPresent(header.getParentHash());
   }
 
   private VoteTally constructMissingCacheEntries(
       final Deque<BlockHeader> headers, final VoteTally tally) {
+    final VoteTally mutableVoteTally = tally.copy();
     while (!headers.isEmpty()) {
       final BlockHeader h = headers.pop();
-      voteTallyUpdater.updateForBlock(h, tally);
-      voteTallyCache.put(h.getHash(), tally.copy());
+      voteTallyUpdater.updateForBlock(h, mutableVoteTally);
+      voteTallyCache.put(h.getHash(), mutableVoteTally.copy());
```

Comment:
For the record, this is the key to the bug fix. Previously the VoteTally that was originally retrieved from the cache was modified and then a copy added for the new block, resulting in the original and new both being the vote tally for the new block.  Now it copies before modifying so the original value is unchanged.